### ISSUE TITLE
Silent payments addresses (BIP352)

### ISSUE THREAD

**Issue by RCasatta** [rust-bitcoin/rust-bitcoin#2778]
I was reading this https://github.com/bitcoindevkit/bdk/issues/1114 and I think implementation should start on rust-bitcoin `Address` struct first

To support sending to silent payment address, at a very high level we need:
* parse an address string like `"sp1qqvvnsd3xnjpmx8hnn2ua0e9sllm34t9jydf8qfesgc7nhdxgzksjwqlrxx37nfzsg6rure5vwa92fksd6f5a6rk05kr07twhd55u3ahquy2v7t6s"`
* convert the silent payment address to a taproot address

[BIP352](https://github.com/bitcoin/bips/blob/master/bip-0352.mediawiki)

**Comment by tcharding** [comment]
Point one can be done, IIUC, by providing a simple public function that calls through to `bech32`. The linked crate uses `bech32 v0.9.0` but the re-write in `v0.11.0` should work too. We could add "first class support" simply by defining a couple of new `bech32::Hrp` consts as we do for the current Bitcoin segwit addresses.

Looks like we could do point 2 in `rust-bitcoin` reasonable simply. I'd love to see this work include integration tests against Core using `bitcoind` - which I've been meaning to get round to for some time.

**Comment by junderw** [comment]
I spent a while the other day thinking of how to fit the concept into BitcoinJS. Here are some musings I came up with if anyone wants to discuss with me.

1. Parsing a sp address is trivial, but converting it to a scriptPubkey requires access to the inputs of the transaction (Psbt) that is being signed.
2. Modifying the input set can potentially invalidate the sp's scriptPubkey. Annoyingly, it also could potentially NOT invalidate it, and also checking for partialSigs and sighashes would also be annoying... I am probably going to set some sort of flag and assume all modification to inputs (insert, delete) will invalidate, it's just more simple this way.
3. There should probably be a new Psbt output data field to store the sp information... If someone imports a Psbt created by someone who had knowledge of the sp address and the second participant decides to add an input, they should know that they need to invalidate and update the output's scriptPubkey. This is especially true for workflows where the creator of Psbts is not a signer, and the signers have the ability to modify inputs etc. sp adoption without this implicitly places the requirement on people who import Psbts they did not create to never believe taproot outputs won't change when they modify inputs.

Anything else I didn't get? I'm all ears... this is a tough one to implement.

**Comment by apoelstra** [comment]
@tcharding 
> We could add "first class support" simply by defining a couple of new `bech32::Hrp` consts as we do for the current Bitcoin segwit addresses.

Yeah, I'd be happy with this, so that people can parse SP addresses into vectors and stuff without needing `expect`s. This should happen in rust-bech32. Maaybe we want a policy about what HRPs we are willing to include there but maybe "Andrew likes this protocol and Clark agrees" is sufficient. Also note that there is an open PR to introduce the crypto into Bitcoin Core https://github.com/bitcoin-core/secp256k1/pull/1519 which is being taken seriously.

There is also a TSP prefix for testnet, from the BIP, and a "SPRT" prefix which appears in the linked crate but not in the BIP text.

From the BIP:

>Note: BIP173 imposes a 90 character limit for Bech32 segwit addresses and limits versions to 0 through 16, whereas a silent payment address requires at least 117 characters[12] and allows versions up to 31. Additionally, since higher versions may add to the data field, it is recommended implementations use a limit of 1023 characters (see BIP173: Checksum design for more details).

In rust-bech32, our bech32(m) checksums have `CODE_LENGTH` set to 1023, which is the design length. However, segwit further imposes this 90-character limit because that's the space that Pieter pseudo-brute-forced and determined he could get better error correcting ability than the code parameters. This limit is enforced in `bech32::segwit::decode` which is what we use in rust-bitcoin for address parsing.

Furthermore, SP does not have this dumb bech32/bech32m switch. It always uses bech32m.

@junderw 
>Parsing a sp address is trivial, but converting it to a scriptPubkey requires access to the inputs of the transaction (Psbt) that is being signed.

Yeah.. furthermore it requires that libsecp PR to be merged. The linked crate hacks together support using the existing rust-secp API but the result is dangerously insecure ... e.g. see [this `calculate_shared_secret` which does a variable-time EC mult])https://github.com/cygnet3/rust-silentpayments/blob/395b153b6d98ea33a59306c1a8a189d4ca152571/src/utils/receiving.rs#L65). (I **think** we could do this better, using the ECDH API rather than mul_tweak, but like, we'd still be doing roll-your-own-crypto in Rust and I don't want to go down that road.).

It would then require basically every existing method on `Address` become fallible, including basic/obvious ones like `script_pubkey`.

**Comment by apoelstra** [comment]
> We should probably also consider adding a BIP 21 URI parser, since I imagine that's how most folks will use silent payments (as it provides backwards compatibility so people who don't support silent payments can still pay you). We'll want something that maps an str to an iterator/Vec, description, and amount.

@TheBlueMatt so, what I would propose is:

* We have a BIP21 crate, which could live in the org (I don't know if we have one already).
* We have a SP crate, which can also live in the org (once it is free of roll-your-own-crypto, which I guess, means, once the libsecp PR is merged and propagated to rust-secp). It depends on bip21, in a feature-gated way.
* Rust-bitcoin exposes both crates' APIs. (This is part of our broader project to make rust-bitcoin a "facade" for all the other functionality.)

**Comment by apoelstra** [comment]
**Meanwhile**, what we can do now, without any new crypto, is:

* Parse SP addresses, validating their length and parsing keys out of them.
* Expose accessors for the raw bytes and/or the keys
* Allow constructing and (re)serializing them from a pair of keys. (And allow generating keys? Unsure if there is value to specialized functions for this beyond the existing keygen functions.)

None of these are hard, but choosing where to put the functionality is. Eventually it should live in the `silentpayments` crate, but that crate is not production-ready, while we can make this basic parsing functionality production-ready. (Though is this useful at all without the heavy crypto? I don't know.) So that's an argument for putting it into rust-bitcoin .. *except* that the existing `silentpayments` crate presumably would want to use these types, and it currently does not have a `bitcoin` dep. And conversely, bitcoin cannot depend on `silentpayments` while it is still experimental.

We could plausibly put it into rust-bech32, which silentpayments *does* depend on, but I don't think that's a sensible home for it. (rust-bech32 does have a whole "segwit" module which we could use as a template, but segwit is much more core to Bitcoin and its parsing logic is way harder to implement without dedicated support from the bech32 lib. SP is basically just "check the length then parse as bech32m".)

**Comment by DanGould** [comment]
> We have a BIP21 crate, which could live in the org (I don't know if we have one already).

@Kixunil's [`bip21` crate](https://github.com/kixunil/bip21) is excellent. It could use a bit more composability and perhaps a builder pattern, but it seems like the best candidate to solve this particular request

> We have a SP crate...It depends on bip21, in a feature-gated way.
> > Shouldn't it be the other way around? BIP 21 should depend on the relevant crates so that it can spit out a list of address structs from the other crates.

I've run into this dependency problem between the payjoin crate and Mutiny Wallet. Kix's bip21 crate has each downstream implementor a generic [`Uri` `Extras`](https://docs.rs/bip21/0.4.0/bip21/struct.Uri.html) type that defines additional parameters beyond the spec. If there were some way to make types that implement this composable, perhaps with a trait, multiple implementations of different `Extras` could be combined so that e.g. `pj` payjoin parameters and `sp` parameters  and bolt12 `lightning` params could be supported together in one bip21 Uri but defined in different places.

**Comment by DanGould** [comment]
I agree wholeheartedly that we need some BIP21 crate that depends on the relevant ecosystem crates that define bitcoin payment instruction formats, and this was what I was actually trying to communicate. We need an interface that allows one to depend on the crates'  implementations and and there doesn't seem to be a good way to do that right now. [Mutiny's Waila crate attempts to support them all](https://github.com/MutinyWallet/bitcoin-waila/blob/master/waila/src/bip21.rs) *without* a connection to underlying dependencies by duplicated implementation.

I think we're digging a layer deeper than the silent payments problem so [I started an issue in the bip21 repo](https://github.com/Kixunil/bip21/issues/22) where I defined the problem. I  want to handle payment instruction formats defined in separate dependencies.

**Comment by Kixunil** [comment]
Only found this conversation now....

I think silent payments are orthogonal to non-silent address especially for the reason that you can't construct script pubkey without additional information. So it'd make sense to me to have it in a separate crate. Though with the recent decision of naming our crate `addresses` it makes even more sense to put it there as a separate type or something. However I'd like the crate to support no-alloc, so it might need to be behind a feature flag.

> I imagine that's how most folks will use silent payments (as it provides backwards compatibility so people who don't support silent payments can still pay you)

I highly doubt this. There are two ways people want to use bitcoin: generate an address every time (in automated systems, to ensure payments can be assigned to senders) or generate it once (for donations or trust-based payments). If you are privacy-conscious enough to use silent payment then you don't really want backwards compatibility as that would mean address reuse. And if you're capable of generating an address for every payment having additional silent payment field in the URI is just noise that takes up QR code space and makes copying the URI more annoying.

However I think using `bitcoin:sp...` makes sense for better UX, so I'd definitely like to update my crate API to support it

> * We have a BIP21 crate, which could live in the org

I'll be happy to transfer my crate to the org.

> and it currently does not have a `bitcoin` dep

Probably needs `primitives` dep or something.

> but I don't think that's a sensible home for it.

I agree. Let `bech32` be just encoding.

> Shouldn't it be the other way around?

No, this is correct.

> We'll maybe also want to have a conversation around LDK making smaller BOLT11+12 crates that BIP 21 could depend on.

I think `lightning-invoice` is fine.

> It could use a bit more composability

It already is composable, just like `serde`. The reasons you think it's not are that you've never seen things composed and there are no helper macros. I definitely want helper macros. The way you compose it is you define `struct MyParams { payjoin: Option<PjParams>, lightning: Option<LnParams> }` (possibly `Option<Result<_, _>>` instead) and make `DeserializationState` contain states of both, you then try to feed the parameters into each and call `finalize` of each at the end. Allowing this is also the reason for the `is_param_known` method.

> I don't buy that some kind of decoder trait handed to `bip21` is a good interface

It is necessary to make it idiomatic Rust - parsing directly into structs, like `serde` does. Only additional helpers are missing.

> We need an interface that allows one to depend on the crates' implementations

It already is there. Just make those crates implement the traits, then compose them in your own struct depending on what you want to support. The same way you'd use `serde`.

**Comment by apoelstra** [comment]
> However I'd like the crate to support no-alloc, so it might need to be behind a feature flag.

Yeah, we would definitely want SP behind a feature flag because it's a big feature and brings in heavy crypto dependencies that don't otherwise exist.

Glad to hear that you have a model for composeability with the existing bip21 API in mind.

**Comment by jlest01** [comment]
> Yeah, we would definitely want SP behind a feature flag because it's a big feature and brings in heavy crypto dependencies that don't otherwise exist.

What additional cryptographic dependencies exactly?
AFAIK, you only need `secp256k1` to create silent payments.

**Comment by Kixunil** [comment]
@jlest01 there's a PR to make the cryptography part of `secp256k1` optional (so just the types stay). You wouldn't need to compile the C library if you don't use SP.

**Comment by apoelstra** [comment]
> What additional cryptographic dependencies exactly?

I mean you're now depending on cryptographic hardness assumptions beyond "signatures are unforgeable" and on cryptographic code which is more complicated and less heavily-vetted than the signature code. Even if it's all present in the same dependency.

**Comment by jlest01** [comment]
@apoelstra thanks for the clarification.

@Kixunil Interesting. Which PR are you referring to ? I couldn't find it.

**Comment by Kixunil** [comment]
https://github.com/rust-bitcoin/rust-secp256k1/pull/703

**Comment by nymius** [comment]
Considering the following comments in this thread or related:

- [**About creating a new silent payment crate**](https://github.com/rust-bitcoin/rust-bitcoin/issues/2778#issuecomment-2118840384)
>We have a SP crate, which can also live in the org (once it is free of roll-your-own-crypto, which I guess, means, once the libsecp PR is merged and propagated to rust-secp). It depends on bip21, in a feature-gated way. 
- [**This idea of supporting parsing, validation and serialization without the silent payment crypto**](https://github.com/rust-bitcoin/rust-bitcoin/issues/2778#issuecomment-2118841880).
> **Meanwhile**, what we can do now, without any new crypto, is:
> 
>     * Parse SP addresses, validating their length and parsing keys out of them.
> 
>     * Expose accessors for the raw bytes and/or the keys
> 
>     * Allow constructing and (re)serializing them from a pair of keys. (And allow generating keys? Unsure if there is value to specialized functions for this beyond the existing keygen functions.)
- [**The necessity of having a common way to interpretate silent payment key information with few dependencies**](https://github.com/rust-bitcoin/bitcoin-payment-instructions/pull/19#issuecomment-3682496259).
> Yea, or even just a subcrate that only does address parsing and none of the other logic, which rust-silentpayments can then depend on. That subcrate would presumably only depend on bech32 or so.

I would like to move this forward.

I have already been working [in address parsing for `bdk-sp`](https://github.com/bitcoindevkit/bdk-sp/blob/master/silentpayments/src/encoding/mod.rs), but with a greater dependency base, the whole `bitcoin` crate. Now that `Network` has been moved to [`bitcoin-network-kind` crate](https://docs.rs/bitcoin-network-kind/latest/bitcoin_network_kind/), it's going to be easier to break this dependency in smaller components.

I will start by moving the referenced address parser to a separate crate/project.

From working in `bdk-sp` I could come up with other structures that may belong to such a crate. My view is that it will contain the encoding/decoding and parsing types and some others used for complementary non cryptographic tasks.

This crate should be rather small and contain a small set of structures:

- `SpCode` (or `SpAddress`, name can be discussed): for encoding/decoding and parsing of silent payment receiver data.
- `SpScan`: key material for services providing scanning to clients.
- `SpSpend`: key material needed to spend silent payment outputs.
- [`SpMeta`](https://github.com/bitcoindevkit/bdk-sp/blob/master/silentpayments/src/receive/mod.rs#L20): complementary data to `TxOut` needed at the moment of spending a silent payment outputs. Targeted for indexing and persistence.
- `Label`: structure to hold information related to silent payment labels. Targeted for indexing and persistence.
- [`LexMin`](https://github.com/bitcoindevkit/bdk-sp/blob/master/silentpayments/src/lib.rs#L83): helper to compute the minimum lexicographic outpoint.

Then any rust project willing to implement applications using silent payments, will have two[^*] upstream dependencies: this "`rust-bip352`" crate and `rust-secp256k1` (compiled with silent payment support).

For BDK, for example, this would allow focusing only on the sourcing, indexing and persistence of the silent payment data, and just connect the components provided by these libraries. At the same time, someone willing to take its own approach, will be free to do it, and still use one or both of the libraries.

cc: @cygnet3 @Sosthene00 @macgyver13
 
[^*]: three with `rust-psbt` if we consider PSBT support, like in BDK

**Comment by Sosthene00** [comment]
> From working in `bdk-sp` I could come up with other structures that may belong to such a crate. My view is that it will contain the encoding/decoding and parsing types and some others used for complementary non cryptographic tasks.
> 
> This crate should be rather small and contain a small set of structures:
> 
>     * `SpCode` (or `SpAddress`, name can be discussed): for encoding/decoding and parsing of silent payment receiver data.
> 
>     * `SpScan`: key material for services providing scanning to clients.
> 
>     * `SpSpend`: key material needed to spend silent payment outputs.
> 
>     * [`SpMeta`](https://github.com/bitcoindevkit/bdk-sp/blob/master/silentpayments/src/receive/mod.rs#L20): complementary data to `TxOut` needed at the moment of spending a silent payment outputs. Targeted for indexing and persistence.
> 
>     * `Label`: structure to hold information related to silent payment labels. Targeted for indexing and persistence.
> 
>     * [`LexMin`](https://github.com/bitcoindevkit/bdk-sp/blob/master/silentpayments/src/lib.rs#L83): helper to compute the minimum lexicographic outpoint.
> 
> 
> Then any rust project willing to implement applications using silent payments, will have two[1](#user-content-fn-*-74362b7e28f6898766ba0932a957ed19) upstream dependencies: this "`rust-bip352`" crate and `rust-secp256k1` (compiled with silent payment support).

Hi Nymius,

What you're describing here looks a lot like what we're trying to do with [spdk](https://github.com/cygnet3/spdk), especially [this branch](https://github.com/cygnet3/spdk/tree/migrate-rust-sp). Slowly converging :)

**Comment by nymius** [comment]
I know, that's why I don't want to put much logic into this crate so `spdk` can do the `scanning for payments from a chain backend, or creating and signing transactions.` by its own, and I stick for it to the BDK lib, at the same time we share common ground.

**Comment by nymius** [comment]
From the discussion above, I came up with the following requirements for the proposed crate:

-  Remove roll-your-own-crypto.
-  Support encoding/decoding.
-  Implement constructors.
-  Support validation.
-  Support serialization/de-serialization.
-  Keep minimal dependencies.
-  Support bip21 (should focus on bip321 now?).
-  Expose raw bytes/keys: I'm not sure what to do with this one. What kind of methods are you looking here?

I've created two branches of the same crate inside `bdk-sp` that fulfill most of these.
The first one: [feat/sp-only-lib](https://github.com/bitcoindevkit/bdk-sp/compare/master...nymius:bdk-sp:feat/sp-code-lib), only exposes `SpCode`. It is targeting encoding/decoding, serialization/de-serialization:

- [x] Do not roll-its-own-crypto.
- [x] Support encoding/decoding.
- [x] Implement constructors.
- [x] Support validation.
- [x] Support serialization/de-serialization.
- [x] Keep minimal dependencies.
- [ ] Support bip21 (should focus on bip321 now?).
- [ ] Expose raw bytes/keys.

The second one: [feat/full-lib](https://github.com/bitcoindevkit/bdk-sp/compare/master...nymius:bdk-sp:feat/full-lib) is integrated with the secp256k1 experimental branch where I've implemented bindings for the silent payment module in libsecp256k1 ([PR#1765](https://github.com/bitcoin-core/secp256k1/pull/1765)):

- [ ] Do not roll-its-own-crypto: Technically is not rolling its own crypto, but depends of a not-yet-merged branch from libsecp256k1.
- [x] Support encoding/decoding.
- [x] Implement constructors.
- [x] Support validation.
- [x] Support serialization/de-serialization.
- [ ] Support bip21 (should focus on bip321 now?).
- [ ] Expose raw bytes/keys.
- [ ] Keep minimal dependencies: it only differs in `bitcoin-primitives` as extra dependency, and requires the `silentpayment` feature from the experimental `rust-secp256k1` branch.

While the first one is simple enough to start using in other projects, the second exposes the silentpayment API to start building on top, and that is desirable for me to keep compatibility between secp256k1 versions. My idea is that `feat/sp-code-lib` will end up being `feat/full-lib` once the moving components like libsecp and the bindings stabilize. I'm planning to keep building on top of the `feat/full-lib` as I progress on the BDK side.

I have yet to research about bip21, bip 321, bip 352, and how `SpCode` fits there but If someone wants to enlighten me, you're welcome.

I would like to hear if this is what you had in mind, or any other feedback.

**Comment by tcharding** [comment]
Hey @nymius I spent some time on this, not exactly sure what you need from me but here are some questions and suggestions that I wrote. Hope this helps.

Notes written reviewing https://github.com/nymius/bdk-sp on both branches linked above. 'this crate' means `bdk-sp`.

### Questions

- Why use `lib/` instead of `bip352`?
- What is the difference between `lib/` and https://github.com/nymius/rust-bip352?
- What is the difference between this crate and these two and why does this crate deserve to be in the org and not the other ones?
   - https://github.com/cygnet3/rust-silentpayments
   - https://github.com/jirijakes/rust-bip352 (has the name on crates.io)

  
- Do we know who `jirijakes` and `cypnet3` are and are they still working on bip352? Are y'all collaborating, competing, or ignoring each other?

### Suggestions

- nit: Use module level docs `//!`
- Could use a bit more CI (could use `cargo rbmt`). Eg multiple toolchains and docs/docsrs builds.
- I wouldn't personally use `Sp` prefix on all the types because this is a silent payments crate, its redundant to say every type is something to do with silent payment.
  

From 30 minutes looking around IMO the code is in a fit state to be brought into the org. After questions above are addressed.

**Comment by cygnet3** [comment]
Hi, I'm still actively working on silent payments-related stuff, but currently mostly working on a wallet ([cygnet3/dana](https://github.com/cygnet3/dana)). Although the wallet currently depends on [rust-silentpayments](https://github.com/cygnet3/rust-silentpayments), this crate was intended to be deprecated once the silent payments module gets merged into libsecp. Just to give some context.

**Comment by nymius** [comment]
Thanks for the review @tcharding , let me address your questions:

> * Why use `lib/` instead of `bip352`?

The `bdk-sp` project focuses on BIP352, so I organized logic into semantic categories like `wallet`, `indexer`, and `oracles`. The `lib` crate re-exposes secp256k1 bindings as cryptographic primitives and types, replacing the `silentpayments` crate, which didn't fit this structure. Using `bip352` would dilute the purpose of other crates, raising questions about whether they are BIP352-specific or general Bitcoin wallet tools.

However, this naming is specific to the experimental `bdk-sp` context. The new crate does not need to retain the `lib` name.

>  * What is the difference between `lib/` and https://github.com/nymius/rust-bip352?

`rust-bip352` was an initial implementation of silent payments, focused solely on cryptographic primitives (with custom crypto). I shifted goals and moved to `bdk-sp`, a more comprehensive project including an integrated wallet and CLI.

Ultimately, silent payments will span multiple crates: primitives in `rust-secp256k1` bindings, non-cryptographic types and PSBT logic here, and indexing, scanning, and sending in BDK.

>  * What is the difference between this crate and these two and why does this crate deserve to be in the org and not the other ones?
>       
>    * https://github.com/cygnet3/rust-silentpayments

`rust-silentpayments` is integrated into `rust-spdk`, a dependency of the [Dana wallet](https://github.com/cygnet3/danawallet), which focuses on silent payments.
We share goals but maintain competing implementations. I continue separately to enable BDK integration and support multiple wallets, aligning with BDK's architecture.
Adopting `spdk`/`Dana` would complicate BDK compatibility.

>    * https://github.com/jirijakes/rust-bip352 (has the name on crates.io)

I am aware of this library but not in contact with the maintainer, so I cannot confirm its status or goals.

Overall, I prioritize maintenance and scrutiny through widespread adoption and multiple reviewers. BDK and `rust-bitcoin` offer stronger positioning than alternatives.

The `lib` crate resembles `spdk`'s address structure but features minimal dependencies, additional tests, and changes emerging from key points of the above discussion. I hope `spdk` benefits from it, while serving other wallets.

> * I wouldn't personally use `Sp` prefix on all the types because this is a silent payments crate, its redundant to say every type is something to do with silent payment.

I considered how types might appear in libraries like BDK handling similar structures. Without context on dependencies, reviewers might overlook silent payments usage. In `rust-secp256k1`, I removed long `Silentpayment` prefixes from functions and types, using aliases and paths instead—e.g., the original `silentpayments_recipient_scan_outputs` from `libsecp256k1` transformed into `sp_rx::scan_outputs`: [here](https://github.com/nymius/rust-secp256k1/blob/52c037bc7fb23d251252553e4898923da4d5d5a3/src/silentpayments/recipient.rs#L555), [here](https://github.com/nymius/rust-secp256k1/blob/52c037bc7fb23d251252553e4898923da4d5d5a3/src/silentpayments/mod.rs#L176), and [here](https://github.com/nymius/rust-secp256k1/blob/52c037bc7fb23d251252553e4898923da4d5d5a3/src/silentpayments/mod.rs#L26).

**Comment by tcharding** [comment]
What bits are you hoping will move into the `rust-bitcoin` org?

> I considered how types might appear in libraries like BDK handling similar structures. 

My personal view, and the view we have taken in `rust-bitcoin`, is that downstream can alias their imports if they need more context e.g., `use foo::bar::Baz as FooBorBaz` and then others who don't need the context don't get the noise and get nice terse names. Also if re-exports are done carefully one can use one level of path on the names eg. 

```rust
fn some_dowstream_function(f: foo::Baz)
```

**Comment by tcharding** [comment]
Do you want me to move this to a discussion so we get threading?
